app/controllers/incomeSources/add/ForeignPropertyCheckDetailsController.scala

```diff
@@ -107,31 +108,58 @@ class ForeignPropertyCheckDetailsController @Inject()(val checkForeignPropertyDe
 
   def getDetails(implicit user: MtdItUser[_]): Future[Either[Throwable, CheckForeignPropertyViewModel]] = {
 
+    sessionService.get(foreignPropertyStartDate).flatMap { startDate: Either[Throwable, Option[String]] =>
+      sessionService.get(addIncomeSourcesAccountingMethod).map { accMethod: Either[Throwable, Option[String]] =>
+        val errors: Seq[String] = getErrors(startDate, accMethod)
+        val result: Option[CheckForeignPropertyViewModel] = getResult(startDate, accMethod)
+
+        result match {
+          case Some(checkForeignPropertyViewModel) =>
+            Right(checkForeignPropertyViewModel)
+          case None =>
+            Left(new IllegalArgumentException(s"Missing required session data: ${errors.map(x => x.mkString(" "))}"))
+        }
+      }
+    }
+
+  }
+
+  def getResult(startDate: Either[Throwable, Option[String]], accMethod: Either[Throwable, Option[String]]): Option[CheckForeignPropertyViewModel] = {
+    (startDate, accMethod) match {
+      case (Right(dateMaybe), Right(methodMaybe)) =>
+        for {
+          foreignPropertyStartDate <- dateMaybe.map(LocalDate.parse)
+          cashOrAccrualsFlag <- methodMaybe
+        } yield {
+          CheckForeignPropertyViewModel(
+            tradingStartDate = foreignPropertyStartDate,
+            cashOrAccrualsFlag = cashOrAccrualsFlag)
+        }
+      case (_, _) => None
+    }
+  }
+
+  def getErrors(startDate: Either[Throwable, Option[String]], accMethod: Either[Throwable, Option[String]]): Seq[String] = {
     case class MissingKey(msg: String)
 
-    val errors: Seq[String] = Seq(
-      user.session.data.get(foreignPropertyStartDate).orElse(Option(MissingKey("MissingKey: addForeignPropertyStartDate"))),
-      user.session.data.get(addIncomeSourcesAccountingMethod).orElse(Option(MissingKey("MissingKey: addIncomeSourcesAccountingMethod")))
+    Seq(
+      startDate match {
+        case Right(nameOpt) => nameOpt match {
```

Comment:
you can simply convert Either to Option as you dropping actual errors provided in Either, 
https://www.scala-lang.org/api/2.13.6/scala/util/Either.html